kratos/python_scripts/analysis_stage.py

```diff
@@ -1,305 +1,376 @@
+        super(DerivedAnalysis, self).__init__(project_parameters)
+
+        Keyword arguments:
+        self -- It signifies an instance of a class.
+        model -- The Model to be used
+        project_parameters -- The ProjectParameters used
+        """
+        if not isinstance(model, KratosMultiphysics.Model):
+            raise Exception("Input is expected to be provided as a Kratos Model object")
+
+        if not isinstance(project_parameters, KratosMultiphysics.Parameters):
+            raise Exception("Input is expected to be provided as a Kratos Parameters object")
+
+        self.model = model
+        self.project_parameters = project_parameters
+
+        ## Get echo level and parallel type
+        self.echo_level = self.project_parameters["problem_data"]["echo_level"].GetInt()
+        self.parallel_type = self.project_parameters["problem_data"]["parallel_type"].GetString()
+        is_distributed_run = KratosMultiphysics.IsDistributedRun()
+
+        if self.parallel_type == "OpenMP" and is_distributed_run:
+            KratosMultiphysics.Logger.PrintWarning("Parallel Type", '"OpenMP" is specified as "parallel_type", but Kratos is running distributed!')
+        if self.parallel_type == "MPI" and not is_distributed_run:
+            KratosMultiphysics.Logger.PrintWarning("Parallel Type", '"MPI" is specified as "parallel_type", but Kratos is not running distributed!')
+
+        self._GetSolver().AddVariables() # this creates the solver and adds the variables
+
+    def Run(self):
+        """This function executes the entire AnalysisStage
+        It can be overridden by derived classes
+        """
+        self.Initialize()
+        self.RunSolutionLoop()
+        self.Finalize()
+
+    def KeepAdvancingSolutionLoop(self):
+        """This function specifies the stopping criteria for breaking the solution loop
+        It can be overridden by derived classes
+        """
+        return self.time < self.end_time
+
+    def RunSolutionLoop(self):
+        """This function executes the solution loop of the AnalysisStage
+        It can be overridden by derived classes
+        """
+        while self.KeepAdvancingSolutionLoop():
+            self.time = self._GetSolver().AdvanceInTime(self.time)
+            self.InitializeSolutionStep()
+            self._GetSolver().Predict()
+            is_converged = self._GetSolver().SolveSolutionStep()
+            self.__CheckIfSolveSolutionStepReturnsAValue(is_converged)
+            self.FinalizeSolutionStep()
+            self.OutputSolutionStep()
+
+    def Initialize(self):
+        """This function initializes the AnalysisStage
+        Usage: It is designed to be called ONCE, BEFORE the execution of the solution-loop
+        This function has to be implemented in deriving classes!
+        """
+        # Modelers:
+        self._CreateModelers()
+        self._ModelersSetupGeometryModel()
+        self._ModelersPrepareGeometryModel()
+        self._ModelersSetupModelPart()

```

Comment:
Now all modeler calls are in separate functions. Do you agree like that @RiccardoRossi ?

Reply:
ok